sterror/sterror.go

```diff
@@ -0,0 +1,157 @@
+package sterror
+
+import (
+	"errors"
+)
+
+type Op string
+
+func (o Op) String() string {
+	return string(o)
+}
+
+type Scope string
+
+func (s Scope) String() string {
+	return string(s)
+}
+
+type Info string
+
+func (i Info) String() string {
+	return string(i)
+}
+
+const (
+	colon     = ": "
+	hyphen    = " - "
+	separator = "\n"
+)
+
+const (
+	Undefined Scope = "Undefined"
+	Host      Scope = "Host"
+	Network   Scope = "Network"
+	Opts      Scope = "Opts"
+	Ospkg     Scope = "Ospkg"
+	Stlog     Scope = "Stlog"
+	Trust     Scope = "Trust"
+)
+
+/* Error-Wrapper
+- op: Function that raised the error
+- Scope: Package that raised the error
+- Info: additional Info passed as a string
+- Err: Wrapped underlying error.
+*/
+type Error struct {
+	Op    Op
+	Scope Scope
+	Info  Info
+	Err   error
+}
+
+// returns the error as a string.
+func (e Error) Error() string {
+	composedErrorString := e.Scope.String()
+
+	switch {
+	case e.Op != "" && e.Info != "":
+		composedErrorString += colon + e.Op.String() + hyphen + e.Info.String()
+	case e.Op != "":
+		composedErrorString += colon + e.Op.String()
+	case e.Info != "":
+		composedErrorString += colon + e.Info.String()
+	}
+
+	if e.Err != nil {
+		composedErrorString += separator + e.Err.Error()
+	}
+
+	return composedErrorString
+}
+
+// create a new Error Wrapper.
+func NewError(args ...interface{}) Error {
+	err := Error{}
+	err.Scope = Undefined
+
+	if len(args) == 0 {
+		return err
+	}
+
+	for _, arg := range args {
+		switch arg := arg.(type) {
+		case Op:
+			err.Op = arg
+		case Scope:
+			err.Scope = arg
+		case Info:
+			err.Info = arg
+		case error:
+			err.Err = arg
+		default:
+		}
+	}
+
+	return err
+}
+
+func Match(got, want Error) bool {
+	if got.Scope != want.Scope {
+		return false
+	}
+
+	if got.Op != want.Op {
+		return false
+	}
+
+	if got.Info != want.Info {
+		return false
+	}
+
+	gotWrappedErr, typeOkGot := got.Err.(Error)
+	wantWrappedErr, typeOkWant := got.Err.(Error)
+
+	if typeOkGot != typeOkWant {
+		return false
+	}
+
+	if typeOkGot {
+		Match(gotWrappedErr, wantWrappedErr)
+	}
+
+	return errors.Is(gotWrappedErr, wantWrappedErr)
+}
+
+// checks if an error is of a specific Scope.
+func IsOfScope(got error, want Scope) bool {
+	gotErr, typeOk := got.(Error)
+	if !typeOk {
+		return false
+	}
+
+	if gotErr.Scope == want {
+		return true
+	} else if gotErr.Err != nil {
+		return IsOfScope(gotErr.Err, want)
+	}
+
+	return false
+}
+
+// checks if an error from a specific operation.
+func IsOfOp(got error, want Op) bool {
+	gotErr, typeOk := got.(Error)
+	if !typeOk {
+		return false
+	}
+
+	if gotErr.Op == want {
+		return true
+	} else if gotErr.Err != nil {
+		return IsOfOp(gotErr.Err, want)
+	}
+
+	return false
+}
```

Comment:
Why do we need those?

Reply:
I thought about the use of both functions for Testing-purposes only.
But you are right, Match() should be enough for testing

Reply:
Put everything you need as test helpers in unexported functions inside the _test.go and place `t.Helper()` at the first line of each such function.